File: wehe/src/main/java/mobi/meddle/wehe/adapter/ImageReplayRecyclerViewAdapter.java
package mobi.meddle.wehe.adapter;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.res.Resources;
import android.net.Uri;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.res.ResourcesCompat;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;
import java.util.Locale;

import mobi.meddle.wehe.R;
import mobi.meddle.wehe.activity.ReplayActivity;
import mobi.meddle.wehe.bean.ApplicationBean;
import mobi.meddle.wehe.constant.Consts;

/**
 * Used to display the apps/ports while the replays are running.
 * XML layout: list_item_replay.xml; Goes in appsRecyclerView in activity_replay.xml
 */
public class ImageReplayRecyclerViewAdapter extends
        RecyclerView.Adapter<ImageReplayRecyclerViewAdapter.ViewHolder> {

    private final List<ApplicationBean> dataList; //list of apps/ports
    private final ReplayActivity replayAct; //replay activity, to get resources
    private final boolean runPortTests; //self explanatory
    private boolean isTomography = false;

    /**
     * Constructor.
     *
     * @param list         list of apps/ports to run replays
     * @param replayAct    a ReplayActivity
     * @param runPortTests true if port tests are being run; false otherwise
     */
    public ImageReplayRecyclerViewAdapter(List<ApplicationBean> list, ReplayActivity replayAct,
                                          boolean runPortTests) {
        this.replayAct = replayAct;
        this.dataList = list;
        this.runPortTests = runPortTests;
    }

    public void setTomography(boolean isTomography) {
        this.isTomography = isTomography;
    }

    @Override
    public int getItemCount() {
        return dataList.size() + 1;
    }

    //runs every time an app/port scrolls onto screen to load that app/port's view
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.xputOriginalTextView.setVisibility(View.GONE);
        holder.xputOriginalValueTextView.setVisibility(View.GONE);
        holder.xputTestTextView.setVisibility(View.GONE);
        holder.xputTestValueTextView.setVisibility(View.GONE);
        holder.arcepLogo.setVisibility(View.GONE);
        holder.alertArcep.setVisibility(View.GONE);
        holder.imageButton.setVisibility(View.GONE);

        if (position == 0) {
            holder.tvAppStatus.setVisibility(View.GONE);
            holder.tvAppTime.setVisibility(View.GONE);
            holder.tvAppSize.setVisibility(View.GONE);
            holder.img.setVisibility(View.GONE);
            String description;
            if (isTomography) {
                description = replayAct.getString(R.string.tomo_test_desc);
            } else if (runPortTests) {
                description = replayAct.getString(R.string.normal_port_desc);
            } else {
                description = replayAct.getString(R.string.normal_app_desc);
            }
            description = ""; //TODO: comment out when time to add test descriptions
            holder.tvAppName.setText(description);
            holder.tvAppName.setTextSize(TypedValue.COMPLEX_UNIT_PX, replayAct.getResources().getDimension(R.dimen.text_medium));
            return;
        } else if (holder.tvAppStatus.getVisibility() == View.GONE) {
            holder.tvAppName.setVisibility(View.VISIBLE);
            holder.tvAppStatus.setVisibility(View.VISIBLE);
            holder.tvAppTime.setVisibility(View.VISIBLE);
            holder.img.setVisibility(View.VISIBLE);
            holder.tvAppName.setTextSize(TypedValue.COMPLEX_UNIT_PX, replayAct.getResources().getDimension(R.dimen.text_small));
        }
        final ApplicationBean app = dataList.get(position - 1);
        Resources res = replayAct.getResources();
        //load name
        holder.tvAppName.setText(app.getName());
        //load time for only apps
        if (runPortTests) {
            holder.tvAppTime.setVisibility(View.GONE);
        } else {
            int time = Consts.TIMEOUT_ENABLED ? Math.min(app.getTime(), Consts.REPLAY_APP_TIMEOUT * 2)
                    : app.getTime();
            holder.tvAppTime.setText(String.format(Locale.getDefault(),
                    res.getString(R.string.replay_time), time));
            holder.tvAppTime.setVisibility(View.VISIBLE);
        }
        //load size
        int numTests = 2 * (app.isTomography() ? Consts.NUM_TOMOGRAPHY_TESTS : 1);
        holder.tvAppSize.setText(String.format(Locale.getDefault(),
                res.getString(R.string.replay_size), numTests, app.getSize()));
        holder.tvAppSize.setVisibility(View.VISIBLE);

        // here we set different color for different results
        int red = res.getColor(R.color.red);
        int green = res.getColor(R.color.forestGreen);
        int yellow = res.getColor(R.color.orange2);
        int blue = res.getColor(R.color.blue0);

        //load arcep alert button and logo
        if (app.getArcepNeedsAlerting()) {
            holder.arcepLogo.setVisibility(View.VISIBLE);
            holder.alertArcep.setVisibility(View.VISIBLE);
            holder.alertArcep.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    app.setArcepNeedsAlerting(false);
                    holder.arcepLogo.setVisibility(View.GONE);
                    holder.alertArcep.setVisibility(View.GONE);

                    //open arcep site in browser; tests will continue running in background
                    Intent i = new Intent(Intent.ACTION_VIEW);
                    i.setData(Uri.parse(Consts.ARCEP_URL));
                    replayAct.startActivity(i);
                }
            });
        }
        //load status
        holder.tvAppStatus.setText(app.getStatus());
        String status = app.getStatus().trim();
        if (status.equals(res.getString(R.string.no_diff))) {
            holder.tvAppStatus.setTextColor(green);
            holder.xputOriginalTextView.setVisibility(View.VISIBLE);
            holder.xputOriginalValueTextView.setVisibility(View.VISIBLE);
            holder.xputTestTextView.setVisibility(View.VISIBLE);
            holder.xputTestValueTextView.setVisibility(View.VISIBLE);
        } else if (status.equals(res.getString(R.string.has_diff))) {
            holder.tvAppStatus.setTextColor(red);
            if (!app.getError().equals(res.getString(R.string.not_all_tcp_sent_text))) {
                holder.xputOriginalTextView.setVisibility(View.VISIBLE);
                holder.xputOriginalValueTextView.setVisibility(View.VISIBLE);
                holder.xputTestTextView.setVisibility(View.VISIBLE);
                holder.xputTestValueTextView.setVisibility(View.VISIBLE);
            }
            holder.imageButton.setVisibility(View.VISIBLE);
            holder.imageButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    makeInfoBox(res.getString(R.string.has_diff), app.getError());
                }
            });
        } else if (status.equals(res.getString(R.string.inconclusive))) {
            holder.tvAppStatus.setTextColor(yellow);
            if (app.getError().equals("")) {
                holder.xputOriginalTextView.setVisibility(View.VISIBLE);
                holder.xputOriginalValueTextView.setVisibility(View.VISIBLE);
                holder.xputTestTextView.setVisibility(View.VISIBLE);
                holder.xputTestValueTextView.setVisibility(View.VISIBLE);
            } else {
                holder.imageButton.setVisibility(View.VISIBLE);
                holder.imageButton.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        makeInfoBox(res.getString(R.string.inconclusive).split(",")[0], app.getError());
                    }
                });
            }
        } else if (status.equals(res.getString(R.string.tomo_failed))) {
            holder.tvAppStatus.setTextColor(red);
            holder.xputOriginalTextView.setVisibility(View.VISIBLE);
        } else if (status.equals(res.getString(R.string.tomo_succ))) {
            holder.tvAppStatus.setTextColor(green);
            holder.xputOriginalTextView.setVisibility(View.VISIBLE);
        } else {
            holder.tvAppStatus.setTextColor(blue);
        }

        //load throughputs
        String appName = runPortTests ?
                String.format(res.getString(R.string.port_name), app.getName().split("\\s+")[1]) : app.getName();
        String port443 = res.getString(R.string.port_name, "443");
        String xputOriginalLabel = res.getString(R.string.xputOriginal, appName);
        String xputTestLabel = runPortTests ? res.getString(R.string.xputOriginal, port443) :
                res.getString(R.string.xputTest, appName);
        String throughputFormat = replayAct.getResources().getString(R.string.throughput);
        String xputOriginal = String.format(Locale.getDefault(), throughputFormat, app.originalThroughput);
        String xputTest = String.format(Locale.getDefault(), throughputFormat, app.randomThroughput);

        if (app.isTomography()) {
            if (app.getDifferentiationNetwork().equals("")) {
                holder.xputOriginalTextView.setText(R.string.tomo_failed_desc);
            } else {
                holder.xputOriginalTextView.setText(String.format(res.getString(R.string.tomo_succ_desc),
                        app.getDifferentiationNetwork()));
            }
        } else {
            holder.xputOriginalTextView.setText(xputOriginalLabel);
            holder.xputTestTextView.setText(xputTestLabel);
            holder.xputOriginalValueTextView.setText(xputOriginal);
            holder.xputTestValueTextView.setText(xputTest);
        }

        holder.img.setImageDrawable(ResourcesCompat.getDrawable(replayAct.getResources(),
                res.getIdentifier(app.getImage(),
                        "drawable", replayAct.getPackageName()), null));
    }

    @NonNull
    @Override
    public ImageReplayRecyclerViewAdapter.ViewHolder onCreateViewHolder(
            @NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(
                R.layout.list_item_replay, parent, false);

        return new ViewHolder(view);
    }

    /**
     * Display an information box if user clicks the "i" icon.
     *
     * @param title title of the info box
     * @param msg   the message to display to the user
     */
    private void makeInfoBox(String title, String msg) {
        new AlertDialog.Builder(replayAct, AlertDialog.THEME_DEVICE_DEFAULT_LIGHT)
                .setTitle(title)
                .setMessage(msg)
                .setNeutralButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which) {
                        // do nothing
                    }
                }).show();
    }

    /**
     * The view that the user sees for each app/port being run in the tests.
     */
    static class ViewHolder extends RecyclerView.ViewHolder {
        // each data item is just a string in this case
        final TextView tvAppName;
        final TextView tvAppSize;
        final TextView tvAppTime;
        final TextView tvAppStatus;
        final TextView xputOriginalTextView;
        final TextView xputOriginalValueTextView;
        final TextView xputTestTextView;
        final TextView xputTestValueTextView;
        final ImageView img;
        final ImageButton imageButton;
        final ImageView arcepLogo;
        final Button alertArcep;

        ViewHolder(View view) {
            super(view);
            xputOriginalTextView = view.findViewById(R.id.xputOriginal);
            xputOriginalValueTextView = view.findViewById(R.id.xputOriginalValue);
            xputTestTextView = view.findViewById(R.id.xputTest);
            xputTestValueTextView = view.findViewById(R.id.xputTestValue);

            tvAppName = view.findViewById(R.id.resultNameText);
            tvAppSize = view.findViewById(R.id.appSize);
            tvAppTime = view.findViewById(R.id.appTime);
            tvAppStatus = view.findViewById(R.id.appStatusTextView);
            img = view.findViewById(R.id.appImageView);
            imageButton = view.findViewById(R.id.ib_reverse_engineer_info);

            arcepLogo = view.findViewById(R.id.arcepLogoImageView);
            alertArcep = view.findViewById(R.id.reportToArcep);
        }
    }
}

File: wehe/src/main/java/mobi/meddle/wehe/adapter/ImageReplayRecyclerViewAdapter.kt
package mobi.meddle.wehe.adapter

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.RecyclerView
import mobi.meddle.wehe.R
import mobi.meddle.wehe.constant.Consts
import mobi.meddle.wehe.data.model.ApplicationBean
import mobi.meddle.wehe.ui.replay.activity.BackgroundReplayActivity
import mobi.meddle.wehe.util.THEME_DEVICE_DEFAULT_LIGHT_DIALOG
import java.util.Locale
import kotlin.math.min

/**
 * Used to display the apps/ports while the replays are running.
 * XML layout: list_item_replay.xml; Goes in appsRecyclerView in activity_replay.xml
 */
class ImageReplayRecyclerViewAdapter(
    private val context: Context,
    private var dataList: List<ApplicationBean>, //replay activity, to get resources
    private val replayAct: BackgroundReplayActivity,
    private val runPortTests: Boolean,
) : RecyclerView.Adapter<ImageReplayRecyclerViewAdapter.ViewHolder>() {
    private var isTomography = false

    fun setTomography(isTomography: Boolean) {
        this.isTomography = isTomography
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        // Return just the actual number of items in the list
        return dataList.size
    }

    // Update the app list
    fun updateApps(newApps: ArrayList<ApplicationBean>) {
        this.dataList = newApps
        notifyDataSetChanged()
    }

    //runs every time an app/port scrolls onto screen to load that app/port's view
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.xputOriginalTextView.visibility = View.GONE
        holder.xputOriginalValueTextView.visibility = View.GONE
        holder.xputTestTextView.visibility = View.GONE
        holder.xputTestValueTextView.visibility = View.GONE
        holder.arcepLogo.visibility = View.GONE
        holder.alertArcep.visibility = View.GONE
        holder.alertFCC.visibility = View.GONE
        holder.imageButton.visibility = View.GONE

        // Make all these views visible by default
        holder.tvAppName.visibility = View.VISIBLE
        holder.tvAppStatus.visibility = View.VISIBLE
        holder.tvAppTime.visibility = View.VISIBLE
        holder.img.visibility = View.VISIBLE

        // Set text size to normal size
        holder.tvAppName.setTextSize(
            TypedValue.COMPLEX_UNIT_PX,
            replayAct.resources.getDimension(R.dimen.text_small)
        )

        val app = dataList[position]
        val res = replayAct.resources
        //load name
        holder.tvAppName.text = app.name
        //load time for only apps
        if (runPortTests) {
            holder.tvAppTime.visibility = View.GONE
        } else {
            val time = if (Consts.TIMEOUT_ENABLED) min(
                app.time.toDouble(),
                (Consts.REPLAY_APP_TIMEOUT * 2).toDouble()
            )
                .toInt() else
                app.time
            holder.tvAppTime.text = String.format(
                Locale.getDefault(),
                res.getString(R.string.replay_time), time
            )
            holder.tvAppTime.visibility = View.VISIBLE
        }
        //load size
        val numTests = 2 * (if (app.isTomography) Consts.NUM_TOMOGRAPHY_TESTS else 1)
        holder.tvAppSize.text = String.format(
            Locale.getDefault(),
            res.getString(R.string.replay_size), numTests, app.size
        )
        holder.tvAppSize.visibility = View.VISIBLE

        // here we set different color for different results
        val red = ContextCompat.getColor(context, R.color.red)
        val green = ContextCompat.getColor(context, R.color.forestGreen)
        val yellow = ContextCompat.getColor(context, R.color.orange2)
        val blue = ContextCompat.getColor(context, R.color.blue0)

        //load arcep alert button and logo
        if (app.arcepNeedsAlerting) {
            holder.arcepLogo.visibility = View.VISIBLE
            holder.alertArcep.visibility = View.VISIBLE
            holder.alertArcep.setOnClickListener {
                app.arcepNeedsAlerting = false
                holder.arcepLogo.visibility = View.GONE
                holder.alertArcep.visibility = View.GONE

                //open arcep site in browser; tests will continue running in background
                val i = Intent(Intent.ACTION_VIEW)
                i.setData(Uri.parse(Consts.ARCEP_URL))
                replayAct.startActivity(i)
            }
        } else if (app.isAlertFCC) {
            holder.alertFCC.visibility = View.VISIBLE
            holder.alertFCC.setOnClickListener {
                app.isAlertFCC = false
                holder.alertFCC.visibility = View.GONE

                //open arcep site in browser; tests will continue running in background
                val i = Intent(Intent.ACTION_VIEW)
                i.setData(Uri.parse(Consts.FCC_URL))
                replayAct.startActivity(i)
            }
        }
        //load status
        holder.tvAppStatus.text = app.status
        val status = app.status.trim { it <= ' ' }
        if (status == res.getString(R.string.no_diff)) {
            holder.tvAppStatus.setTextColor(green)
            holder.xputOriginalTextView.visibility = View.VISIBLE
            holder.xputOriginalValueTextView.visibility = View.VISIBLE
            holder.xputTestTextView.visibility = View.VISIBLE
            holder.xputTestValueTextView.visibility = View.VISIBLE
        } else if (status == res.getString(R.string.has_diff)) {
            holder.tvAppStatus.setTextColor(red)
            if (app.error != res.getString(R.string.not_all_tcp_sent_text)) {
                holder.xputOriginalTextView.visibility = View.VISIBLE
                holder.xputOriginalValueTextView.visibility = View.VISIBLE
                holder.xputTestTextView.visibility = View.VISIBLE
                holder.xputTestValueTextView.visibility = View.VISIBLE
            }
            holder.imageButton.visibility = View.VISIBLE
            holder.imageButton.setOnClickListener {
                makeInfoBox(
                    res.getString(R.string.has_diff),
                    app.error
                )
            }
        } else if (status == res.getString(R.string.inconclusive)) {
            holder.tvAppStatus.setTextColor(yellow)
            if (app.error == "") {
                holder.xputOriginalTextView.visibility = View.VISIBLE
                holder.xputOriginalValueTextView.visibility = View.VISIBLE
                holder.xputTestTextView.visibility = View.VISIBLE
                holder.xputTestValueTextView.visibility = View.VISIBLE
            } else {
                holder.imageButton.visibility = View.VISIBLE
                holder.imageButton.setOnClickListener {
                    makeInfoBox(
                        res.getString(R.string.inconclusive).split(",".toRegex())
                            .dropLastWhile { it.isEmpty() }.toTypedArray()[0],
                        app.error
                    )
                }
            }
        } else if (status == res.getString(R.string.tomo_failed)) {
            holder.tvAppStatus.setTextColor(red)
            holder.xputOriginalTextView.visibility = View.VISIBLE
        } else if (status == res.getString(R.string.tomo_succ)) {
            holder.tvAppStatus.setTextColor(green)
            holder.xputOriginalTextView.visibility = View.VISIBLE
        } else {
            holder.tvAppStatus.setTextColor(blue)
        }

        //load throughputs
        val appName = if (runPortTests) String.format(res.getString(R.string.port_name),
            app.name.split("\\s+".toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()[1]) else app.name
        val port443 = res.getString(R.string.port_name, "443")
        val xputOriginalLabel = res.getString(R.string.xputOriginal, appName)
        val xputTestLabel = if (runPortTests) res.getString(
            R.string.xputOriginal,
            port443
        ) else res.getString(R.string.xputTest, appName)
        val throughputFormat = replayAct.resources.getString(R.string.throughput)
        val xputOriginal =
            String.format(Locale.getDefault(), throughputFormat, app.originalThroughput)
        val xputTest = String.format(Locale.getDefault(), throughputFormat, app.randomThroughput)

        if (app.isTomography) {
            if (app.differentiationNetwork == "") {
                holder.xputOriginalTextView.setText(R.string.tomo_failed_desc)
            } else {
                holder.xputOriginalTextView.text = String.format(
                    res.getString(R.string.tomo_succ_desc),
                    app.differentiationNetwork
                )
            }
        } else {
            holder.xputOriginalTextView.text = xputOriginalLabel
            holder.xputTestTextView.text = xputTestLabel
            holder.xputOriginalValueTextView.text = xputOriginal
            holder.xputTestValueTextView.text = xputTest
        }

        holder.img.setImageDrawable(
            ResourcesCompat.getDrawable(
                replayAct.resources,
                res.getIdentifier(
                    app.image,
                    "drawable", replayAct.packageName
                ), null
            )
        )
    }

    override fun onCreateViewHolder(
        parent: ViewGroup, viewType: Int
    ): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(
            R.layout.list_item_replay, parent, false
        )

        return ViewHolder(view)
    }

    /**
     * Display an information box if user clicks the "i" icon.
     *
     * @param title title of the info box
     * @param msg   the message to display to the user
     */
    private fun makeInfoBox(title: String, msg: String) {
        AlertDialog.Builder(replayAct, THEME_DEVICE_DEFAULT_LIGHT_DIALOG)
            .setTitle(title)
            .setMessage(msg)
            .setNeutralButton(
                android.R.string.ok
            ) { _, _ ->
                // do nothing
            }.show()
    }

    /**
     * The view that the user sees for each app/port being run in the tests.
     */
    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        // each data item is just a string in this case
        val tvAppName: TextView = view.findViewById(R.id.resultNameText)
        val tvAppSize: TextView = view.findViewById(R.id.appSize)
        val tvAppTime: TextView = view.findViewById(R.id.appTime)
        val tvAppStatus: TextView = view.findViewById(R.id.appStatusTextView)
        val xputOriginalTextView: TextView = view.findViewById(R.id.xputOriginal)
        val xputOriginalValueTextView: TextView =
            view.findViewById(R.id.xputOriginalValue)
        val xputTestTextView: TextView = view.findViewById(R.id.xputTest)
        val xputTestValueTextView: TextView = view.findViewById(R.id.xputTestValue)
        val img: ImageView = view.findViewById(R.id.appImageView)
        val imageButton: ImageButton = view.findViewById(R.id.ib_reverse_engineer_info)
        val arcepLogo: ImageView = view.findViewById(R.id.arcepLogoImageView)
        val alertArcep: Button = view.findViewById(R.id.reportToArcep)
        val alertFCC: Button = view.findViewById(R.id.alertFcc)
    }
}